Add content around working with non-sufficient assets (#774)
polkadot-developers/polkadot-docs: develop/interoperability/best-practices-for-teleporting-assets.md

!!!note Sufficient Assets
    Always check the chain to determine what assets are considered sufficient.
    This may vary from chain to chain.

### Non-Sufficient Assets

Non-sufficient assets can:

- Pay for transaction and XCM execution/delivery fees by swapping for a sufficient asset.
- Be used to create new accounts by swapping for an exact amount of a sufficient asset to cover the existential deposit of the new account.

!!!note Swapping Non-Sufficient Assets
    Swapping non-sufficient assets for a sufficient asset can be done with the help of [asset conversion](https://wiki.polkadot.network/learn/learn-asset-conversion-assethub){target=\_blank}, which is live on Asset Hub. Always ensure that there is an associated liquidity pool with healthy liquidity for the pair that you intend to swap.

Always verify that the amount expected to be received in the recipient account is the amount expected after all fees are deducted (transaction fees, XCM fees, and/or swapping fees).

From a UX perspective, when designing cross-chain applications, consider who will be paying for the fees.

> **Example scenario**: Sending 100 USDT to a new Asset Hub account.

Reviewer: I didn't realize these were the same. Could be worth doing a text snippet. Don't want to keep adding to this PR though, but something to consider

Author: I removed the one under polkadot-protocol/architecture/system-chains/asset-hub.md as the design suggestion is not specific to Asset Hub. It is better to have it one place under best practices.

- Option 1 - Deduct from sending amount:

    - **Receiver gets**: 100 USDT - 0.01 DOT (for ED) - transaction and XCM fees.
    - **Sender pays**: Exactly 100 USDT total.
    - Simpler for the sender, but the receiver gets less than expected.

- Option 2 - Additional sender cost:

>     - **Receiver gets**: full 100 USDT + 0.01 DOT (for ED).

Reviewer: ```suggestion
    - **Receiver gets**: Full 100 USDT + 0.01 DOT (for ED).
```

One last thing

## Conclusion

Successful asset teleportation requires meticulous validation, proper error handling, and comprehensive testing. Always prioritize safety over convenience, implement thorough dry run testing, and maintain robust monitoring systems to ensure reliable cross-chain asset transfers.

The key to preventing asset loss is comprehensive pre-flight validation combined with proper XCM construction and execution monitoring. When in doubt, always dry run transactions on both source and destination chains before execution.

## Where to Go Next

<div class="grid cards" markdown>

-   <span class="badge guide">Tutorial</span> __Batch Teleport__

    ---

    A tutorial detailing the step-by-step process of batch teleporting assets using the ParaSpell SDK.

>     [:octicons-arrow-right-24: Batch Teleport](/tutorials/polkadot-sdk/system-chains/asset-hub/batch-teleport-assets){target=\_blank}

Reviewer: ```suggestion
    [:octicons-arrow-right-24: Batch Teleport](/tutorials/polkadot-sdk/system-chains/asset-hub/batch-teleport-assets/){target=\_blank}
```

---
title: Best Practices for Teleporting Assets
description: An in-depth guide on best practices for teleporting assets using XCM
tutorial_badge: Intermediate
---

# Best Practices for Teleporting Assets

## Introduction

Cross-chain asset transfers in the Polkadot ecosystem require careful planning and validation to ensure successful execution. Unlike simple blockchain transactions that either succeed or fail cleanly, XCM operations can result in trapped or lost assets if not properly constructed.

This comprehensive guide outlines essential best practices for teleporting assets using XCM, from pre-flight checks to transaction execution. You'll learn how to avoid common pitfalls, implement proper validation, and ensure your cross-chain transfers succeed reliably.

## Prerequisites

Before implementing asset teleports, ensure you have:

- Strong knowledge of [XCM (Cross-Consensus Messaging)](/develop/interoperability/intro-to-xcm/){target=_blank}
- Firm understanding of [sufficient and non-sufficient assets](/polkadot-protocol/architecture/system-chains/asset-hub/#sufficient-and-non-sufficient-assets){target=_blank}

> - Understanding of [existential deposits](/polkadot-protocol/glossary/#existential-deposit){target=_blank} and the [Substrate account model](/polkadot-protocol/parachain-basics/accounts/){target=_blank}

Reviewer 2: Throughout the documentation, we tend to use Polkadot SDK instead of Substrate. Not sure if it’d be better here, but here is my attempt:

```suggestion
- Understanding of the concept of [existential deposits](/polkadot-protocol/glossary/#existential-deposit){target=\_blank} and the [Polkadot SDK account model](/polkadot-protocol/parachain-basics/accounts/){target=\_blank}
```

## Teleporting an Asset

Asset teleportation requires comprehensive pre-flight validation to prevent failures and asset loss. 

Here is an example flow of teleporting an asset from a parachain to Asset Hub.

> ???- example "Teleporting a Non-Sufficient Asset from a Parachain to Asset Hub"

Reviewer 2: I'm not sure if `???-` is the correct syntax for collapsible; in the docs, we use a green icon. If I'm not mistaken, the syntax should be just: `???` 

Please check other instances

    ```mermaid
    flowchart TD

>         A[Teleporting a <br/>Non-Sufficient Asset<br/>from a Parachain to <br/>Asset Hub] --> B[Check account existence<br/>on Asset Hub]

Reviewer 2: first step overlaps with the title above

Reviewer 2: I mean, you can directly start with the "Check account existence" step, since that's the real action. "Teleporting a non-sufficient asset" is not a concrete step itself.

        B --> C{Account exists?}
        C -->|No| D[Account needs to be <br/>created with sufficient <br/>asset for ED]

>         C -->|Yes| E[Check existential deposit<br/>on Asset Hub]

Reviewer 2: Something I don't understand is why, if the account exists, I still need to check the existential deposit?

Author: Yes technically if an account exists then it either has an existential deposit or a reference tied to it. I have this here so that the dev not only checks of an account address exists but also checks that it has an ED.

        
        D --> F[Ensure sufficient asset<br/>available for ED]
        F --> E
        
        E --> G{Has existential deposit?}
        G -->|No| H[Need sufficient asset<br/>for ED requirement]
        G -->|Yes| I[Check fee coverage]
        
        H --> J[Swap non-sufficient asset<br/>for sufficient asset via<br/>Asset Conversion]
        J --> I
        
        I --> K{Enough funds for fees?}
        K -->|No| L[Estimate fees<br/>using Transaction Payment <br/>API and XCM Payment API]
        K -->|Yes| M[Dry-run cross-chain <br/>transaction]
        
        L --> N[Estimate fee in <br/>non-sufficient asset<br/>using Asset Conversion API]
        N --> O[Ensure sufficient balance<br/>for fees + transfer amount]
        O --> M
        
        M --> P[Local dry-run on <br/>source chain]
        P --> Q{Local dry-run success?}

>         Q -->|No| R[Fix local issues<br/>- Insufficient balance<br/>- Invalid XCM construction]

Reviewer 2: ```suggestion
        Q -->|No| R[Fix local issues, e.g.:<br/>- Insufficient balance<br/>- Invalid XCM construction]
```

        Q -->|Yes| S[Remote dry-run on <br/>Asset Hub]
        
        R --> P
        
        S --> T{Remote dry-run success?}
        T -->|No| U[Common Issues:<br/>- Missing ED<br/>- Invalid account<br/>- Fee payment failure]
        T -->|Yes| V[Execute cross-chain <br/>transaction]
        
        U --> W[Modify XCM e.g. include<br/>ExchangeAsset instruction<br/>to swap for ED]
        W --> S
        
        V --> X[Transaction complete<br/>asset transferred <br/>successfully]
        
        style A fill:#e1f5fe
        style X fill:#e8f5e8
        style R fill:#ffebee
        style U fill:#ffebee
        style J fill:#fff3e0

>         style W fill:#fff3e0

Reviewer 2: Please check with the regulators about these styles, unsure if they overwrite something previously defined according to the polkadot brand kit.

    ```

## Account Existence Verification

Always verify that the destination account exists or can be created before initiating transfers. Account queries should check:

- Account existence on the destination chain

> - Current balance and nonce information  

Reviewer 2: The `nonce` value needs to be checked here as well?

Author: I think it is more of a verification that the transaction succeeded so if the account has a nonce of 1 and after the transaction it has a nonce of 2, then that is an extra confirmation that the transaction succeeded. But if after the transaction is continues with a nonce of 1 then something went wrong...

- Whether the account meets existential deposit requirements

> ???- example "Account Existence Check Examples"

Reviewer 2: I'd add here in the middle an introductory sentence like: "You can refer to the following snippets in case you want to check ed of a specific account:" or something like that, feel free to paraphrase it

>     **Using PolkadotJS API:**

Reviewer 2: I suggest using tabs here for the sake of shortening the page a bit, for references, check -> https://github.com/polkadot-developers/polkadot-docs/blob/master/develop/toolkit/api-libraries/dedot.md?plain=1#L32-L45

It renders as - https://docs.polkadot.com/develop/toolkit/api-libraries/dedot/#installation

      const existentialDeposit = api.consts.balances.existentialDeposit.toBigInt();
      
      return {
        exists: !accountInfo.isEmpty,
        balance,
        hasExistentialDeposit: balance >= existentialDeposit
      };
    }
    ```

    **Using Polkadot API (PAPI):**
    ```typescript
    import { dot } from "@polkadot-api/descriptors";
    import { createClient } from "polkadot-api";
    import { getWsProvider } from "polkadot-api/ws-provider/web";

    async function queryAccountBalance(address: string) {
      const client = createClient(getWsProvider("wss://rpc.polkadot.io"));
      const api = client.getTypedApi(dot);
      
      const accountInfo = await api.query.System.Account.getValue(address);
      const existentialDeposit = await api.constants.Balances.ExistentialDeposit();
      
      return {
        balance: accountInfo?.data.free ?? 0n,
        hasED: (accountInfo?.data.free ?? 0n) >= existentialDeposit
      };
    }
    ```

### Existential Deposit Considerations

Existential deposits vary by network and must be maintained to keep accounts active:

- **Polkadot**: 1 DOT (10^10 planck)
- **Kusama**: 0.0033 KSM (3.3 * 10^10 planck)  

> - **Asset Hub**: 0.1 DOT (10^9 planck)

Reviewer 2: What do you think about moving this to a note?

> For non-sufficient assets, ensure the destination account has sufficient native token or sufficient asset to maintain the ED, or include asset conversion instructions in the XCM.

Reviewer 2: ```suggestion
For non-sufficient assets, make sure the destination account has enough native tokens or assets to maintain the ED; otherwise, include asset conversion instructions in the XCM.
```

## Fee Estimation and Coverage

### Runtime API Integration

> Use proper runtime APIs for accurate fee estimation rather than hardcoded values:

Reviewer 2: Nitpicking here, but I'm doubting if this should be `fee` or `fees` (same for the title). Any thoughts?

> Use proper runtime APIs for accurate fee estimation rather than hardcoded values:

Reviewer 2: what about refactoring this to be a list instead, like:
- Transaction Payment API
    // snippet
- XCM Payment API
    // snippet
- Asset Conversion API
    // snippet

> [Transaction Payment API](https://paritytech.github.io/polkadot-sdk/master/pallet_transaction_payment_rpc/trait.TransactionPaymentRuntimeApi.html){target=\_blank} for local fees:

Reviewer 2: ```suggestion
[Transaction Payment API](https://paritytech.github.io/polkadot-sdk/master/pallet_transaction_payment_rpc/trait.TransactionPaymentRuntimeApi.html){target=\_blank} for local fees:

```

```typescript
import { dot } from "@polkadot-api/descriptors";
import { createClient } from "polkadot-api";
import { getWsProvider } from "polkadot-api/ws-provider/web";

const client = createClient(getWsProvider("wss://rpc.polkadot.io"));
const api = client.getTypedApi(dot);

const feeDetails = await api.apis.TransactionPaymentApi.query_fee_details(
  call.toHex(),
  call.encodedLength
);
```

> [XCM Payment API](https://paritytech.github.io/polkadot-sdk/master/xcm_runtime_apis/fees/trait.XcmPaymentApi.html){target=\_blank} for cross-chain delivery fees:

Reviewer 2: ```suggestion
[XCM Payment API](https://paritytech.github.io/polkadot-sdk/master/xcm_runtime_apis/fees/trait.XcmPaymentApi.html){target=\_blank} for cross-chain delivery fees:

```

```typescript
import { dot } from "@polkadot-api/descriptors";
import { createClient } from "polkadot-api";
import { getWsProvider } from "polkadot-api/ws-provider/web";

const client = createClient(getWsProvider("wss://rpc.polkadot.io"));
const api = client.getTypedApi(dot);

const deliveryFees = await api.apis.XcmPaymentApi.query_delivery_fees(
  destination,
  message
);
```

> [Asset Conversion API](https://paritytech.github.io/polkadot-sdk/master/pallet_asset_conversion/trait.AssetConversionApi.html){target=\_blank} (available on Asset Hub) for fee conversion:

Reviewer 2: ```suggestion
[Asset Conversion API](https://paritytech.github.io/polkadot-sdk/master/pallet_asset_conversion/trait.AssetConversionApi.html){target=\_blank} (available on Asset Hub) for fee conversion:

```

```typescript
import { ah } from "@polkadot-api/descriptors";
import { createClient } from "polkadot-api";
import { getWsProvider } from "polkadot-api/ws-provider/web";

const client = createClient(getWsProvider("wss://rpc.polkadot.io"));
const api = client.getTypedApi(ah);

const feeInAsset = await api.apis.AssetConversionApi.quote_price_exact_tokens_for_tokens(
  assetIn,
  assetOut,
  amountIn
);
```

### Multi-Asset Fee Handling

> For non-sufficient assets, implement fee payment strategies:

Reviewer 2: ```suggestion
For dealing with non-sufficient assets and fees, there are different fee payment strategies you can implement:
```

1. **[Asset Conversion](/polkadot-protocol/architecture/system-chains/asset-hub/#non-sufficient-assets){target=\_blank}**: Convert part of the transfer amount to pay fees
2. **Separate Fee Payment**: Use a different sufficient asset for fees
3. **Fee Sponsorship**: Have another account pay fees on behalf of the user

## Asset Type Considerations

### Sufficient Assets

> !!!note Sufficient Assets

Reviewer 2: I'd move this note to the end of this sub-section

    Always check the chain to see what assets are considered sufficient assets.
    This may vary from chain to chain.

[Sufficient assets](/polkadot-protocol/architecture/system-chains/asset-hub/#sufficient-assets){target=\_blank} can:

- Pay for transaction fees directly
- Meet existential deposit requirements and can suffice for creating new accounts

**Best Practices:**

- Verify minimum transfer amounts meet destination requirements
- Account for both transaction and existential deposit costs
- Handle automatic account creation scenarios

### Non-Sufficient Assets

[Non-sufficient assets](/polkadot-protocol/architecture/system-chains/asset-hub/#non-sufficient-assets){target=\_blank} require:

- Existing destination accounts with ED
- Alternative fee payment mechanisms

**Best Practices:**

- Always verify recipient account status
- Include asset conversion for ED if needed
- Plan for fee payment using sufficient assets

> - Test with realistic scenarios including account creation

Reviewer 2: Consider reorganizing this into a table to reduce the length of the page.

## Comprehensive Dry Run Testing

> Dry run testing represents the most critical step in preventing asset loss during teleportation. Execute dry runs on both the source and destination chains to validate every aspect of the transfer before committing to the actual transaction. Local dry runs on the source chain validate transaction construction correctness, sufficient balance for transfer and fees, and proper XCM message generation. Remote dry runs on the destination chain verify that the XCM message will execute successfully, including asset reception and processing, account creation or balance updates, and proper fee deduction. This two-phase validation approach catches the majority of potential issues before they can cause problems in production.

Reviewer 2: This info is very valuable, I'd try to highlight it doing something like:
```suggestion
Dry run testing represents the most critical step in preventing asset loss during teleportation. Execute dry runs on both the source and destination chains to validate every aspect of the transfer before committing to the actual transaction. You can perform local dry runs or remote dry runs:

- Local dry runs on the source chain validate transaction construction correctness, sufficient balance for transfer and fees, and proper XCM message generation. 
- Remote dry runs on the destination chain verify that the XCM message will execute successfully, including asset reception and processing, account creation or balance updates, and proper fee deduction.

This two-phase validation approach catches the majority of potential issues before they can cause problems in production.
```

### Example Dry Run Implementation

???- example "Comprehensive Dry Run Example"

    ```typescript title="dry-run-example.ts"

>     --8<-- 'code/develop/interoperability/best-practices-for-teleporting-assets/dry-run-example.ts:0:180'

Reviewer 2: If you want to show the entire snippet, you can simply remove the numbers, and the pre-processor will fetch the whole file.

    ```

## Common Errors and Prevention

Asset teleportation failures can occur for various reasons, each with specific causes and prevention strategies. Understanding these common error patterns helps you implement proper validation and avoid the most frequent pitfalls that lead to failed transfers or trapped assets. Each error type requires different diagnostic approaches and preventive measures to ensure reliable cross-chain operations.

### Common Account and Balance Issues

> **[FailedToTransactAsset](https://paritytech.github.io/polkadot-sdk/master/cumulus_primitives_core/enum.XcmError.html#variant.FailedToTransactAsset) Errors**

Author: Not really.

> **[FailedToTransactAsset](https://paritytech.github.io/polkadot-sdk/master/cumulus_primitives_core/enum.XcmError.html#variant.FailedToTransactAsset) Errors**

Reviewer: Then maybe we should just list the error in the title instead (this goes for all the sections under `## Common Errors and Prevention`)

`### FailedToTransactAsset`
